File: week5/python/IsGraphBipartite_785.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        
        # colors = [0] * len(graph)
        colors = dict()
        def dfs(node: int, color: int) -> bool:
            if node in colors:
                return colors[node] == color
            colors[node] = color
            for child in graph[node]:
                if not dfs(child, -color):
                    return False
                    
            return True
        
        for node in range(len(graph)):
            if node not in colors and not dfs(node, 1):
                return False
            
        return True

File: week5/java/IsGraphBipartite_785.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        if (graph == null || graph.length < 2)
            return true;
        
        int[] colors = new int[graph.length];
        for (int i = 0; i < graph.length; i++) {
            if (colors[i] != 0)
                continue;
            if (!colorGraph(graph, i, 1, colors))
                return false;
        }
        return true;
    }
    
    private boolean colorGraph(int[][] graph, int index, int color, int[] colors) {
        if (colors[index] != 0)
            return colors[index] == color;
        
        colors[index] = color;
        for (int i : graph[index]) {
            if (!colorGraph(graph, i, -color, colors))
                return false;
        }
        return true;
    }
}
